### ISSUE TITLE
labeled-response: what is a response?

### ISSUE THREAD

**Issue by syzop** [ircv3/ircv3-specifications#379]
I can see how the labeled-response draft can be helpful to client developers, so I've been looking into it and it raises a number of questions... one of which is very basic: what is a response?

For something like ```WHOIS Nick``` the response should be labeled. That's clear and what I expect.

Apparently receiving a ```PRIVMSG``` is not a reponse, as the spec says:
> When a client sends a message to itself, the server MUST NOT include the label tag, except for any acknowledgment sent with the echo-message mechanism. This is because the delivered message is not a response, but a side effect.

Which makes... some sense...... I guess.

But how does this expand to other things?
* KICKing someone
* Setting a TOPIC
* JOINing a channel
* KILLing a user

What are the rules? When is something a response and when is it "generated traffic" that should not have a labeled response?

* Are all numeric replies considered responses and anything else is not? Apparently in https://github.com/ircv3/ircv3-ideas/issues/37 @csmith expects ```PONG``` to be considered a response (which I can understand), so then no.. not only numeric replies. <- XXX seems I misread what he wrote, it was a server oriented PING, question still stands though.. if a client originated PING would require the PONG to be labeled, if yes.. then why is this different than PRIVMSG. If no, why is PONG not a response to a PING?

* Is all traffic directed to yourself considered a response? No, the ```PRIVMSG``` example to self (without echo) explicitly says this is not the case

There must be some clear rule here that both servers and clients can follow and rely on. So what is it? ;)

**Comment by syzop** [comment]
Perhaps worth mentioning what I thought about before having read the full specification.

Implementation rule:
* Any line sent back to the originating client caused while executing the command will include the labeled-response label

So that means that, for example ```JOIN #chan``` would cause the following whole blob to be included in the same labeled-response batch:
* The join itself :xxx JOIN ...
* TOPIC (numerics)
* NAMES list (numerics)
* Basically any traffic sent back to the client caused by the single "JOIN #chan" command

Similarly if the user does "KICK #chan someone" the KICK would be seen with the label.

In extreme cases the batch could be quite large, for example an IRCOp issueing a SQUIT would result in a BATCH that includes the entire netsplit including quits and so on (that would even be a nested batch with labeled-response and netsplit).

And so on and so on.

It would have been a simple rule and for any command it would be clear what would be considered a response.

But.. apparently that is not the idea. Which is fine by me if you think that is better (maybe share "why" though). But that does bring me back to the opening post :)

**Comment by jwheare** [comment]
> But.. apparently that is not the idea

Where do you get this from? Your understanding sounds correct for me. The privmsg example is consistent with it.

**Comment by DanielOaks** [comment]
> _Any line sent back to the originating client caused while executing the command will include the labeled-response label_

This is the rule I stick by and how I implement this in Oragono. The paragraph and all mentioned above basically mean _"Don't forward the label on to anyone other than the originating client"_ as it's not useful to anyone but the client who originated the command and would just confuse them.

For reference, what I do on executing a command is create a small 'buffer' for the response for lines that are sent back to the originating client, and depending on how many lines are sent back to that client (and what capabilities they support), it'll send it back as a batch, as a single line with the label tag, or create the empty message/batch so the client knows there were no reply lines. Works really nicely, and means I don't need to think about it.

**Comment by syzop** [comment]
It seems @jwheare and I have some disagreement, he is saying that a PRIVMSG to self (and this is also what the spec says) should not have the labeled response. So I am like: why is it special? He says: because it is not a response, the target just happens to be the source also. (Feel free to correct me James)

I think it starts to be confusing if you do things like that, by that reasoning you could also say the ":xyz KICK #chan zzz" line is a broadcast and not a response and should not have the label. And then you can have doubts about the JOIN part as well where the numerics would have the label but the :xxx JOIN will not, and so on and so on.

I would just rather just stick to the general rule as outlined in my comment, and - if I understand correctly - what DanielOaks says as well.

**Comment by DanielOaks** [comment]
So that PRIVMSG thing is a special case. It's basically in effect if you have `echo-message` enabled.

This is the correct behaviour with `echo-message` and label, when you send a message to yourself:
```
C: @label=abc PRIVMSG dan :hi myself!
S: @label=abc :dan!user@host PRIVMSG dan :hi myself!   <- this is the `echo-message` response, the response from executing the `PRIVMSG` commang
S: :dan!user@host PRIVMSG dan :hi myself!              <- this is the server actually sending the msg
```

And if you were sending to a channel or to another user:
```
# what 'dan' sees sending the message
C: @label=abc PRIVMSG #somechan :oh hi!
S: @label=abc :dan!user@host #somechan :oh hi!   <- this is the `echo-message` response

# what others see in the channel
S: :dan!user@host PRIVMSG #somechan :oh hi!      <- this is the server actually sending the msg
```

It's just saying that on the `echo-message` response, include the label (since that's the output from executing the command), and then when you actually send the message on, the receiving client shouldn't see the `label` tag at all. It's a little confusing but hopefully those examples lay it out a bit.

**Comment by syzop** [comment]
Yeah I don't mean echo-message, best to keep that out of the conversation.

**Comment by syzop** [comment]
The issue is that James considers (without echo-message) a privmsg to yourself a special case where the standard rule does not apply and the label must be stripped.
I am wondering if we start doing it there, then what happens next and question the logic behind it.
It seems James thinks it is only limited to that special case, though, I should mention that.

**Comment by DanielOaks** [comment]
fwiw I agree with James' reading of it/behaviour, and how the spec describes it is how Ora implements it when echo-message isn't enabled as well.

**Comment by jwheare** [comment]
Summary after chatting with @syzop in IRC:

Basically, only the privmsg to self needs to be implemented carefully. Whether it's a "special case" or not depends on your architecture, but it just needs to be consistent for clients to be able to properly deal with the double message situation when echo-message is on.

In any other case, the server architecture may vary on whether a KICK message seen by the initiator is a confirmation/acknowledgment/response or a side-effect/broadcast message. I personally feel that this doesn't matter too much, as long as the "failure" case is that a label is NOT sent out. It would be more of an issue if it failed by potentially leaking a label to other clients.

So, yeah implement this as @DanielOaks says, or according to the implementation rule mentioned by @syzop (with a special case for the self privmsg behaviour if necessary) and it should be simple enough.

Make sense to everyone? Does this match your implementation in InspIRCD too @SaberUK?

As a follow up, do we need to add some clarifying language to the spec around this?

**Comment by SaberUK** [comment]
>Does this match your implementation in InspIRCD too @SaberUK?

As far as I know yes.

**Comment by DanielOaks** [comment]
Probably couldn't hurt to have a closer look at the language. I know that we had a very similar convo about this as we were writing the spec up, and probably won't be the last. All else fails, some more specific examples with commands that have both a response and a side-effect showing what to do in each case could help explain it.

**Comment by jwheare** [comment]
@syzop any wording in mind that might have avoided you having to open this issue?

**Comment by syzop** [comment]
@jwheare I think you should explain in the specification why you consider PRIVMSG (and presumably NOTICE) to self a special case and not the other examples I gave. As you know, on IRC you did not convince me, so.. I would not be in a position to enhance the text in that respect since I don't agree with your reasoning ;)

In addition to that (but not only): yes, @DanielOaks, I certainly think adding an example with an user issued JOIN would be helpful.

**Comment by jwheare** [comment]
Huh, I thought I had convinced you. Here was my explanation of an example of why it's important with messages sent to yourself:

```
19:10:12 <jwheare> right but in the case of self privmsg it's also bad because clients can't
then treat those buffers consistently
19:13:52 <jwheare> e.g. in my client i allow people to pm themselves as like a notepad thing.
so when label and echo is on, i want to prevent them getting double messages. i do that by
simply dropping any self messages that have a label
19:14:24 <~Syzop[AWAY]> :)
19:14:31 <~Syzop[AWAY]> yeah I understand
19:14:33 <jwheare> so if you include label on all self messages, even the "side effect" ones,
against the spec, my client will no longer show any messages i send to myself
19:14:42 <~Syzop[AWAY]> right
```

I took that "right" and "I understand" to mean that you understand why it's important to special case it.

You seem to be worried about an imagined special case slippery slope, but there isn't one. It's important that you treat privmsg and notice that way, but not really anything else.

> It seems James thinks it is only limited to that special case, though, I should mention that. And if that is true, then I couldn't care less, but if it is not true then... it is bad.

That *is* true, so what is still left for you to be convinced?

**Comment by syzop** [comment]
Oh that must be a misunderstanding then. I understood that you, in your implementation, relied on certain behavior. That was it.

You never convinced me that a PRIVMSG to self, which one could call traffic and not a reply, would be different than for example the ```:xyz JOIN #chan``` or ```:xyz KICK``` or other messages. PRIVMSG carries a message, KICK carries a kick, etc., why would PRIVMSG be special. They are not replies but events or traffic by your logic.
Indeed in our conversation you mentioned that there was a point in seeing it as such and leaving it up to the server implementation to label or not to label these traffic messages.

Anyway, now that I am re-reading the conversation in this thread, are you saying that the only reason that PRIVMSG is exempt is because of the echo-request issue? If the echo-request feature never existed, there would be no need to give PRIVMSG special treatment? I mean *then* I can understand. Then it's simply working around an issue you created with echo-request.

**Comment by jwheare** [comment]
Yes it’s because of echo-message.

**Comment by syzop** [comment]
Ok, so this is all a workaround for echo-message.

So then it's about this text:

> When a client sends a message to itself, the server MUST NOT include the label tag, except for any acknowledgment sent with the echo-message mechanism. **This is because the delivered message is not a response, but a side effect.**

First of all, it is not clear to which of the cases "This" refers (the former or the latter), but more importantly it talks about the delivered message "not as a response but a side effect". Which is how all the confusion started. It raised the question and this entire thread as to what is a response and what is a side effect or traffic or whatever.

I suggest something like:

> When a client sends a message to itself, the server MUST NOT include the label tag, except for any acknowledgment sent with the echo-message mechanism. This because otherwise the client is unable to differentiate between an echo-message and the actual message to self.

**Comment by jwheare** [comment]
Sure, that makes sense.

**Comment by syzop** [comment]
Since this is a workaround for an issue created by ```echo-message```, I think it should be swapped: the echo-message should not include the label, but the other message should. Then everything makes sense. For a client with echo-request capability it does not make any difference.

So:

> When a client sends a message to itself, the server MUST NOT include the label tag in the acknowledgment sent with the echo-message mechanism. This because otherwise the client is unable to see the difference between an echo-message and an actual message to itself.

**Comment by syzop** [comment]
Whoops, retry:

Since this special case is an attempt to workaround an issue created by ```echo-message``` I wonder if it wouldn't be better to drop the label in the echo-request in case of a message to self. Then it all makes sense.

For clients with echo-message there would be no difference in behavior compared to what has been proposed before (they still get 1 message with and 1 message without the label). And for clients without echo-message it would just label everything (no strange exception with self msgs).

So:
> When a client sends a message to itself, the server MUST NOT include the label tag for the acknowledgment sent with the echo-message mechanism. This because otherwise the client is unable to see the difference between an echo-message and an actual message to itself.

(or some better wording)

**Comment by jwheare** [comment]
I don't think it makes sense to label the received message instead of the acknowledgement.

**Comment by syzop** [comment]
As we analyzed in this thread the workaround only exists due to ```echo-message``` causing an effect that the message to self and the echo are indistinguishable.

If, for example, echo-message would have added an ```echo``` label to indicate the echoed message then this issue would not have occurred. Then it would have been like this:

With echo-message but without labeled-response:
```
@echo :self!user@host PRIVMSG self :hello!
:self!user@host PRIVMSG self :hello!
```

With echo-message and labeled-response:
```
@echo,label=xxxx :self!user@host PRIVMSG self :hello!
:@label=xxxx :self!user@host PRIVMSG self :hello!
```

Then the whole workaround would not have been required. And similar if a different command was chosen. Reusing it was a mistake and is being rectified by a workaround in the labeled-response draft specification.

Not criticism, the past is the past and we can only improve. This is simply to show the reasoning and why things were chosen as they were.

So, since the issue is echo-message here, it only makes sense to apply the workaround in case the ```echo-message``` CAP is requested.

**Comment by jwheare** [comment]
Perhaps it would help if you explained why this makes it easier for your implementation while the proposed solution is harder?

**Comment by syzop** [comment]
Let me put it another way, or shorter, in any case:

Snce the workaround is due to ```echo-message``` it should only affect ```echo-message```. Perhaps another text suggestion:

> When a client sends a message to itself and the echo-message capability
is enabled, then the server MUST NOT include the label tag [....]

Fill in at [..] what you want... drop the label in echoed one or the non-echoed one, just choose one, I don't mind. Client won't see the difference anyway.

**Comment by jwheare** [comment]
This is addressed in #385
